### PR TITLE
Overlay Bulk Lookup

### PR DESCRIPTION
**Pull request by jenlij**
This change is [Reviewable](https://reviewable.io/reviews/storj/storj/279)

### CODE REVIEW

File: pkg/overlay/cache.go [hunk 1]
```diff
@@ -82,6 +82,31 @@ func (o *Cache) Get(ctx context.Context, key string) (*overlay.Node, error) {
...
+	var ns []*overlay.Node
+	for _, v := range vs {
+		na := &overlay.Node{}
+		if v == nil {
+			ns = append(ns, nil)
+		} else if err := proto.Unmarshal(v, na); err == nil {
+			ns = append(ns, na)
+		} else {
+			//or we could append a nil value rather than returning an error. @team What do you think?
+			return nil, OverlayError.New("could not unmarshal non nil node: %s", err)
```
Comment (jtolds) [inline pkg/overlay/cache.go]: re: your comment, i think this is the right behavior, but i think this err value isn't the err value from the earlier if, is it? i'm confused about the scoping.

either way, might be more clear to do something like:

```
for _, v := range vs {
  if v == nil {
    ns = append(ns, nil)
    continue
  }
  na := &overlay.Node{}
  err := proto.Unmarshal(v, na)
  if err != nil {
    return nil, OverlayError.New("could not unmarshal non-nil node: %v", err)
  }
  ns = append(ns, na)
}
```

File: storage/redis/client.go [hunk 2]
```diff
@@ -117,11 +117,16 @@ func (client *Client) GetAll(keys storage.Keys) (storage.Values, error) {
 
 	values := []storage.Value{}
 	for _, result := range results {
-		s, ok := result.(string)
-		if !ok {
-			return nil, Error.New("invalid result type %T", result)
+		if result == nil {
```
Comment (egonelbre) [inline storage/redis/client.go]: It's unclear why this change is necessary here.

File: storage/boltdb/client.go [hunk 3]
```diff
@@ -125,11 +125,14 @@ func (client *Client) GetAll(keys storage.Keys) (storage.Values, error) {
 	err := client.view(func(bucket *bolt.Bucket) error {
 		for _, key := range keys {
 			val := bucket.Get([]byte(key))
-			vals = append(vals, storage.CloneValue(storage.Value(val)))
+			if val == nil {
```
Comment (egonelbre) [inline storage/boltdb/client.go]: It's unclear why this change is necessary here.

Comment (egonelbre) [reply storage/boltdb/client.go]: Also indent the error flow rather than keeping both if / else. See https://github.com/golang/go/wiki/CodeReviewComments#indent-error-flow

File: pkg/utils/utils.go [hunk 4]
```diff
@@ -62,3 +64,22 @@ func (errs combinedError) Error() string {
 	}
 	return ""
 }
+
+//LookupRequestsToNodeIDs returns the nodeIDs from the LookupRequests
+func LookupRequestsToNodeIDs(reqs *proto.LookupRequests) []string {
```
Comment (egonelbre) [inline pkg/utils/utils.go]: Is there a better place where these methods could be?

File: pkg/overlay/client_test.go [hunk 5]
```diff
@@ -128,6 +128,41 @@ func TestLookup(t *testing.T) {
 	}
 
 }
+func TestBulkLookup(t *testing.T) {
+	cases := []struct {
```
Comment (egonelbre) [inline pkg/overlay/client_test.go]: There's a single case here so no need for tables. Also rather than test internal behavior, test for expected external behavior.

Comment (jenlij) [reply pkg/overlay/client_test.go]: What do you mean by internal vs external behavior?

File: pkg/utils/utils.go (hunk 4 above)
Comment (jenlij) [reply pkg/utils/utils.go]: We had thought they may be useful as helpers later on, but right now they're only being used in overlay/server.go. So I could move them in there if you think that's better style.

File: storage/boltdb/client.go (hunk 3 above)
Comment (jenlij) [reply storage/boltdb/client.go]: Ok, interesting - good to know, I didn't know that was a preferred style. Thx

File: storage/redis/client.go (hunk 2 above)
Comment (jenlij) [reply storage/redis/client.go]: I want to make sure I append the nil value to the results returned by GetAll and the original implementation will append an empty storage value, rather than actual nil.

File: pkg/utils/utils.go (hunk 4 above)
Comment (egonelbre) [reply pkg/utils/utils.go]: If they are used only in overlay/server.go, then yes... I would also consider unexporting them until actually needed to make public.

File: storage/redis/client.go (hunk 2 above)
Comment (egonelbre) [reply storage/redis/client.go]: I don't think the underlying redis nor bolt implementations give that guarantee. This unfortunately means that everything above them should be able to deal with the `nil == storage.Value{}` situation.

Comment (egonelbre) [reply storage/redis/client.go]: Although, the change here is correct for another reason. i.e. GetAll shouldn't fail on a missing value.

File: pkg/overlay/client_test.go (hunk 5 above)
Comment (egonelbre) [reply pkg/overlay/client_test.go]: The internal behavior is how the thing under the test achieves it's outcome. The external behavior is about what is expected out of the thing under test.

Testing internal behaviors is usually fragile and low-risk, meaning there's little to gain from testing it. Similarly, internal behavior can change and break tests, while all code is still correct. Let's say you introduce a caching layer into `client` then the bulk lookup may not be called at all.

External behavior is usually higher risk. In this case, there are few things that should hold for BulkLookup.

1. The behavior under missing ids, invalid ids.
2. Behavior on `BulkLookup([]dht.NodeID{})`
3. Behavior on randomized ordering of ids. That you do get things back in the right order or mapping.
2.1. Insert test-data in random order
2.2. add missing and invalid id's your initial order
2.3. Repeat n times:
2.4.1. shuffle the items
2.4.2. call GetAll with the first `rand.Intn(n)` elements
2.4.3. check whether the result is the right one

As for the table vs none, there's always many possibilities in extensions, however most of the time possible extensions end up unused.

File: storage/redis/client.go (hunk 2 above)
Comment (jenlij) [reply storage/redis/client.go]: It looks like they should return nil if the key is not found...
https://redis.io/commands/mget
https://godoc.org/github.com/boltdb/bolt#Bucket.Get

File: pkg/utils/utils.go (hunk 4 above)
Comment (jenlij) [reply pkg/utils/utils.go]: sounds good

File: storage/redis/client.go (hunk 2 above)
Comment (egonelbre) [reply storage/redis/client.go]: What do they return for an empty value?

Comment (jenlij) [reply storage/redis/client.go]: empty string for key returns nil, "ERR wrong number of arguments for 'mget' command" if no keys are passed in. I will do a check for that, so the program doesn't panic.

File: pkg/overlay/client_test.go [hunk 6]
```diff
@@ -127,7 +130,147 @@ func TestLookup(t *testing.T) {
...
+			responses: []*proto.Node{nil, nil},
+			errors:    nil,
+		},
+		{testID: "random order and nil",
+			nodeIDs: func() []dht.NodeID {
+				id1 := kademlia.StringToNodeID("n1")
+				id2 := kademlia.StringToNodeID("n2")
+				id3 := kademlia.StringToNodeID("n3")
+				id4 := kademlia.StringToNodeID("n4")
+				return []dht.NodeID{id2, id1, id3, id4}
```
Comment (egonelbre) [inline pkg/overlay/client_test.go]: This is not random order.

File: pkg/overlay/client_test.go [hunk 7]
```diff
@@ -127,7 +130,147 @@ func TestLookup(t *testing.T) {
...
+	if err != nil {
+		return nil, nil, err
+	}
+	identOpt, err := identity.ServerOption()
+	if err != nil {
+		return nil, nil, err
+	}
+
+	grpcServer := grpc.NewServer(identOpt)
+	cache, err := NewRedisOverlayCache("127.0.0.1:6379", "", 1, nil)
```
Comment (egonelbre) [inline pkg/overlay/client_test.go]: Don't assume people have redis installed, use `redisserver` package.

File: pkg/overlay/overlay_test.go [hunk 8]
```diff
@@ -62,3 +63,28 @@ func TestOverlayLookup(t *testing.T) {
 	assert.NoError(t, err)
 	assert.NotNil(t, r)
 }
+
+func TestOverlayBulkLookup(t *testing.T) {
+	lis, err := net.Listen("tcp", fmt.Sprintf(":%d", 0))
```
Comment (egonelbre) [inline pkg/overlay/overlay_test.go]: Start test-servers on `127.0.0.1:0`, other ports may be blocked or other hosts may trigger firewall.
